src/models/geograph.py

```diff
@@ -614,21 +669,317 @@ def get_graph_components(
+            # same patches in the main df
+            self.df: gpd.GeoDataFrame = data
+            if (
+                name is not None
+                and graph is not None
+                and valid_classes is not None
+                and barrier_classes is not None
+                and max_travel_distance is not None
+            ):
+                self.name: str = name
+                self.graph: nx.Graph = graph
+                self.valid_classes: List[Union[str, int]] = valid_classes
+                self.barrier_classes: List[Union[str, int]] = barrier_classes
+                self.max_travel_distance: float = max_travel_distance
+                self.add_distance: bool = add_distance
+        elif isinstance(data, (str, os.PathLike)):
+            load_path = pathlib.Path(data)
+            assert load_path.exists()
+            # Load from saved graph
+            if load_path.suffix in (".pickle", ".pkl", ".gz", ".bz2"):
+                self._load_from_graph_path(load_path)
+            else:
+                raise ValueError(
+                    f"""Extension {load_path.suffix} unknown.
+                                 Must be one of .pickle, .pkl, .gz, .bz2."""
+                )
+        else:
+            raise ValueError(
+                """Type of `data` unknown. Must be a dataframe or file path."""
+            )
+        print("Calculating components...")
+        self.components = self.get_graph_components(
+            calc_polygons=True, add_distance_edges=add_component_edges
+        )
+        self.metrics: Dict[str, Metric] = {}
+        print(
+            f"Habitat successfully loaded with {self.graph.number_of_nodes()} nodes",
+            f"and {self.graph.number_of_edges()} edges.",
+        )
+
+    def _load_from_graph_path(self, load_path: pathlib.Path) -> None:
+        """
+        Load networkx graph and dataframe objects from a pickle file.
+
+        Args:
+            load_path (pathlib.Path): Path to a pickle file. Can be compressed
+            with gzip or bz2.
+
+        Returns:
+            gpd.GeoDataFrame: The dataframe containing polygon objects.
+        """
+        if load_path.suffix == ".bz2":
+            with bz2.BZ2File(load_path, "rb") as bz2_file:
+                data = pickle.load(bz2_file)
+        elif load_path.suffix == ".gz":
+            with gzip.GzipFile(load_path, "rb") as gz_file:
+                data = pickle.loads(gz_file.read())
+        else:
+            with open(load_path, "rb") as file:
+                data = pickle.load(file)
+        self.df = data["dataframe"]
+        self.name = data["name"]
+        self.valid_classes = data["valid_classes"]
+        self.barrier_classes = data["barrier_classes"]
+        self.max_travel_distance = data["max_travel_distance"]
+        self.add_distance = data["add_distance"]
+
+    def save_habitat(self, save_path: pathlib.Path) -> None:
+        """
```

Comment:
For my understanding: how is this different from the main save method of the `GeoGraph` class?

Reply:
I guess the only difference is the `data = { ... } ` part?

Reply:
This also saves all the metadata for the habitat, such as the list of valid classes & barrier classes, the `max_travel_distance`, etc. And the loading function loads all that stuff as well.